### PR TITLE
Use an explicit value mapping instead of IRReplaceSizes()

### PR DESCRIPTION
**Pull request by tlemo**
The core of this change is in lower_validation.cpp & fusion.cpp : instead of mutating the IR during PrepareForLowering(), we build a Val* -> NamedScalar* map, which is part of the Fusion object and used during kernel code generation.

This simplifies the lowering logic, avoids a significant number of IR node allocations and  importantly it allows a simple model for expression evaluation: things work the same before and after lowering (so this is fixing #87)

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/index_compute.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: Please leave this check for now as it's not checked inside the merge/split functions. If we did not compute an index it is a hard error, and helps prevent me from shooting myself in the foot. If we get to a point where the checks cause high latency, or we have a bullet proof codegen then we can get rid of it.

File: torch/csrc/jit/codegen/cuda/fusion.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion.cpp:217]: Would it make sense in our printers to add an output/warning if nothing will be printed like "No value map entries found."

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion.h:222]: What's the significance if we look for a value in the lowered map and don't find one?

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp:221]: Will we have to add this to every all other types?

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_loops.cpp:64]: Why bother when we're guaranteed to set it to something below?
If it's triggering, why not:
```
Val* size = alloc_dims.size() == 0 ? new Int(1) : alloc_dims[0];
```

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp:187]: That's fair, not sure why I insist on using the range constructor, though I have been consistently.

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_validation.cpp:30]: why?

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_validation.cpp:99]: Why?

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_validation.cpp:102]: why?

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_validation.cpp:122]: Why?

**Review by csarofeen — COMMENTED** [review]
There's a lot of stylistic changes in the PR, some fine, some I think are a bit questionable. Stylistic changes are fine to adhere to convention, but not at decreased readability. You'll see my potentially obnoxious comments below asking for clarity on some changes, while others that I didn't comment seem fine.

The general approach seems reasonable to unblock other work, however, I'm curious what other options you considered before going with a val map. I'm concerned about general usage of it moving forward. Are we still in agreement that we would like a hard separation between the user IR and a lowered  IR with its own container? If so, I think it would be nice to put some TODO's or comments that we will be unlikely to continue with the value map approach if so. I'm particularly concerned about potential confusion when using this map/working with it as if someone arbitrarily uses it to store some information, you have to explicitly look for that information. I'm thinking someone may store information in here in a lowering pass, and then someone working on a later pass wouldn't know to look for that information in this map.

File: torch/csrc/jit/codegen/cuda/fusion.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/fusion.cpp:217]: Do you mean, make it obvious that there's nothing to print? The code below does that by bracing the values between the header/footer lines. Empty map would look like:

```
Values map
-------------------------
-------------------------
```

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/fusion.h:222]: Good question. The intention is to apply this mapping unconditionally, for all values during lowering. This mean that if a value is not mapped to something else we should return the original value. Does it make sense?

File: torch/csrc/jit/codegen/cuda/index_compute.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: The check is not gone (index_map_.at(id) is checked access). We do lose the developer friendly message, if you think that's important here I'll revert this change.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp:221]: Right now we only map Ints (sizes), so this would be the most likely point of failure. I didn't want to spray this paranoid check everywhere, although I could be easily be convinced to do so.

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/lower_loops.cpp:64]: 1. Consistency. It builds muscle memory, for best effect this should be a pattern where one doesn't have to think about twice.

2. The practice guards no only against current code, but also future modifications.

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:187]: If I had to venture a guess, it's probably because std::vector<T> is not covariant, so in many cases you really have to use the range constructor. But fortunately, not in this case.

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/lower_validation.cpp:30]: Why was it needed in the first place?

This is the only place we do it, so it's inconsistent. Besides, are you confident that future modifications would correctly duplicate the conditions in the comments?

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/lower_validation.cpp:102]: just a minor simplification (I think I rewrote more of this code, then reverted most of it)

If you don't like it I'll revert this change.

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/lower_validation.cpp:122]: The main style that we seem to follow prescribes that abbreviations and acronyms to be spelled as regular words (StartRpc() rather than StartRPC()) : https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

I'm fine with deviating from this style if we document it and do it consistently (although my recommendation is to stick to the well documented Google style - if anyone has any stylistic questions we can point them there)

**Comment by tlemo** [comment]
Re. the general direction, I agree, there are many things I don't like about the mapping approach, but I think it's good as a stop gap. (the fact that the IR is not heavily mutated during lowering is an important step towards a clear lowering boundary)

I'd love to hear any suggestions on how we can make sure this mapping is not getting abused in the meantime.

File: torch/csrc/jit/codegen/cuda/fusion.cpp
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/fusion.cpp:217]: I was thinking a small message like "Warning: No values in map" between the lines would be a bit more intuitive.

File: torch/csrc/jit/codegen/cuda/fusion.h
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/fusion.h:222]: Sorry, this became more clear when I got later in the PR, didn't review my review before sending.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: I always forget about at. That makes sense, can you put the error message in a comment before the at call so we have a breadcrumb for ourselves.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp:221]: Given that this is a stopgap, yes, that makes sense.

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_validation.cpp:102]: On second review the auto on the for loop argument is fine as the name pretty strictly implies it's a TensorView. The auto on the root_td I dislike, as the naming is conflicting with types. getRootDomain explicitly does not return a TensorDomain and it's important to keep in mind that this type is a vector of IterDomains, not a TensorDomain.

Generally I do not prefer auto over stricter typing and myself should use auto more sparingly. Many people still don't use modern IDEs and figuring out types can be painful.

File: test/cpp/jit/test_gpu.cpp
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp:794]: How does something like this seem to you?
https://github.com/csarofeen/pytorch/blob/manual/test/cpp/jit/test_gpu.cpp#L3264-L3359

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/lower_validation.cpp:99]: reverted

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/lower_validation.cpp:102]: reverted

File: torch/csrc/jit/codegen/cuda/index_compute.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: done

**Review by csarofeen — APPROVED** [review]
Looks good, feel free to merge.

File: test/cpp/jit/test_gpu.cpp
Comment (tlemo) [reply test/cpp/jit/test_gpu.cpp:794]: It could work. Looks a bit too fuzzy to me ("foo bar" would compare equal to "foobar"), so it would be nice if it collapses spaces between identifiers (to a single space) instead of removing them completely, but in this case it may be good enough.

File: torch/csrc/jit/codegen/cuda/lower_validation.cpp
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/lower_validation.cpp:102]: > The auto on the root_td I dislike, as the naming is conflicting with types. getRootDomain explicitly does not return a TensorDomain and it's important to keep in mind that this type is a vector of IterDomains, not a TensorDomain.

This is a good point. Would it make sense to rename getRootDomain() (and TendorDomain::rootDomain())? Or make them return TensorDomain?

**Pull request merged**
